jep/0000/README.adoc

```diff
@@ -0,0 +1,298 @@
+
+=== Stapler SPI
+
+Stapler defines a `DispatchValidator` SPI to offload most of the dispatchability logic to that implementation.
+Similar to JEP-218, this was done to make changes to Stapler fairly contained and allow for more rapid iterations of development.
+This allows for Stapler to remain backward compatible in its behavior while allowing Jenkins to move forward with a safer web framework.
+Stapler also defines a `ScriptExecutor` interface to reduce duplication of dispatch validation logic between different facets.
+
+=== Separate Annotation Library
+
+To allow plugin developers to make use of the new and updated annotations, the previously created `jenkins-stapler-support` annotation library is updated with `@StaplerViews` and `@StaplerFragments`.
+This allows plugins to use the annotations without requiring a recent release of Jenkins as a dependency.
+
+=== Lack of JSP and JRuby Support
+
+Stapler supports views written in other languages besides Jelly and Groovy.
+However, the vast majority of views in Jenkins and Jenkins plugins are written in those two languages.
+The existing facets and dispatchers for JSP and ERB (JRuby) views do not support view dispatch filtering and will instead continue to be dispatched as before.
+
+=== Rejection of Fragment Filename Syntax Approach
+
+One proposed way to indicate that a view is a fragment was to prefix fragment view filenames with an underscore.
+For example, a view named `configure.jelly` may include a descriptor list of `_config.jelly` fragments.
+Transparent rewrites of includes to fragment views would preserve compatibility between plugins.
+This idea was rejected eventually due to combinatorical compatibility complexities.
+
+=== Rejection of META-INF Service Approach
+
+One proposed way to address this was by using the `META-INF/services/` service discovery mechanism to contribute filters.
+This was rejected as it complicates overriding default behavior.
+Instead, filters that wish to extend other filters can do so explicitly by creating a composite filter wrapper.
+
+=== Rejection of 404 Not Found Approach
+
+During development, view dispatch was cancelled by responding with a 404 Not Found error.
+This approach would prevent further dispatchers such as `doDynamic` from operating along with any other dispatcher with lower precedence than view dispatch.
+
+=== Rejection of Standard Request Attribute Approach
+
+A fairly similar implementation to the one described in this JEP was previously developed with more of the validation business logic exposed in Stapler.
+This standardized a servlet request attribute used for storing the dispatchability status of the request along with providing a more limited SPI for offloading some of the validation logic to Jenkins.
+This approach was rejected after significant refactoring that allowed for moving the business logic of how requests track their validation status down to Jenkins which offered better implementation flexibility.
```

Comment:
This sentence is a little unclear, particularly the part "allowed for moving the business logic of how requests track".

I don't immediately have good alternative proposals, though. I'd have to think it through more.